### ISSUE TITLE
Nonparametric tests: Kruskal-Wallis test

### ISSUE THREAD

**Issue by ashokkrish** [ashokkrish/CougarStats#4]
Under Statistical Inference when the user selects:

<img width="292" alt="image" src="https://github.com/user-attachments/assets/1e2026eb-7617-4f20-80a1-ff4f24fb000f" />

we should have a radio button toggle like this

<img width="261" alt="image" src="https://github.com/ashokkrish/CougarStats/assets/410949/c74fd99b-5440-4edd-ae2e-f73e841f6a38"> 

Whether the user selects **ANOVA** or **Kruskal-Wallis** in both cases we will have the option to upload a dataset

Upload your Data (.csv or .xls or .xlsx or .txt)
[ BROWSE ] _____________________________________

[Calculate] [Reset] 

If the user selects **Kruskal-Wallis** then on the mainPanel we will have the following workflow.

**Kruskal-Wallis Test Workflow**

**Step 1** Setting up the null and alternate hypothesis: The null hypothesis for the Kruskal-Wallis test is that there are no differences between the medians of the groups, meaning that all the groups are sampled from populations with the same distribution.

The alternate hypothesis, on the other hand, states that there is at least one group whose median is different from the medians of the other groups. In other words, it suggests that there are differences in the medians of at least one pair of groups being compared.

**Step 2** Fix _α_.

**Step 3** Instead of the ANOVA table we need a table of ranks. We covered Kruskal-Wallis tests in MATH 2444. I have posted some notes and examples [here](https://drive.google.com/drive/u/1/folders/1lgPNdtw1xWTSkig64S4YIYiEtO_VRgZU). Grab the TS-value from the `kruskal.wallis()` R function. The test statistic formula is

<img width="190" alt="image" src="https://github.com/ashokkrish/CougarStats/assets/410949/c457a239-d9eb-42bb-9f79-c013790ac101">

The next line after printing the formula grab the numbers from the table of ranks and substitute here.

**Step 4** The critical value or P-value method.

**CRITICAL-VALUE APPROACH**

The critical value is dependent on the degrees of freedom which can be calculate as follows.

Next draw the right-skewed Chi-squared distribution curve, identify AR and shade the RR. The test is right-sided so shade on the right side tail only. Locate the TS-value on the curve in red colour

**P-VALUE APPROACH**

Grab the p-value from the `kruskal.wallis()` R function and compare it against _α_.

**Step 5** Conclusion: If the null hypothesis is rejected then we conclude that at least two medians differ.

Sample R code for the **Kruskal-Wallis Test** to run command line

```
# Sample data
Diatoms <- c(425, 466, 440, 498)
Bacteria <- c(303, 301, 293, 328, 327, 314)
Macroalgae <- c(327, 324, 302, 312, 303)

# Kruskal-Wallis test
result <- kruskal.test(list(Diatoms, Bacteria, Macroalgae))

# Calculate the chi-square critical value for the degrees of freedom
df <- result$parameter
crit_val <- qchisq(0.05, df, lower.tail = FALSE)

# Print the results
cat("Kruskal-Wallis test:\n")
cat("Test statistic:", result$statistic, "\n")
cat("Degrees of freedom:", df, "\n")
cat("P-value:", result$p.value, "\n")
cat("Critical value at alpha = 0.05:", crit_val, "\n")
```

**Comment by bryce-carson** [comment]
@ashokkrish, I'm just letting you know that no notification was sent to me when you edited this because the tag (i.e., `@bryce-carson`) was written improperly (`@ bryce`).

**Comment by bryce-carson** [comment]
@ashokkrish, what is the source of data for this test? Is it the uploaded data provided by the fileInput widget used for the ANOVA methodology? Is that the only other widget that should be included in this methodology?
```
[ ] Inference about 1 sample
[ ] Inference about 2 samples
[X] Inference about more than 2 samples (e.g. ANOVA or Kruskal-Wallis)
[ ] Inference for Categorical Data (e.g χ2 test)

[X] Kruskal-Wallis
[ ] ANOVA

Upload your Data (.csv or .xls or .xlsx or .txt)
[ BROWSE ] _____________________________________

[Calculate] [Reset]
```

**Comment by ashokkrish** [comment]
@bryce-carson > what is the source of data for this test?

We are going to have the user to upload the data the same way they did for ANOVA. After the upload is successful have this radioButton toggle

<img width="271" alt="image" src="https://github.com/user-attachments/assets/ca9ebb67-a79a-4c52-b1f2-c7eada84a14c">

The mainPanel will have the following three tabPanels

<img width="256" alt="image" src="https://github.com/user-attachments/assets/9633e2d6-c1e0-493f-b739-005bc1dbc5a9">

I would prefer if the uploaded data immediately gets displayed on the mainPanel even before anything else. The same applies to ANOVA.

**Comment by ashokkrish** [comment]
@sam-vb Could start working on this task please?

**Comment by ashokkrish** [comment]
@sam-vb Sorry I forgot to mention this, for both ANOVA and the K-W test we need to have this field after the file upload widget

<img width="132" alt="image" src="https://github.com/user-attachments/assets/a36df0b8-9c3b-4876-8ebf-6de05a2903e9" />

**Comment by sam-vb** [comment]
@ashokkrish File upload is partially implemented (will still do validations, likely the same way MMyer did so with ANOVA), and I will add the significance level - are there any other elements you'd like included in the side panel? If not, do you have any visual samples of what the main panel should look like?

**Comment by ashokkrish** [comment]
@sam-vb You can refer to [this](https://drive.google.com/drive/folders/1lgPNdtw1xWTSkig64S4YIYiEtO_VRgZU?usp=drive_link) Google drive folder for some examples. On the main panel we are going display the five-step process of hypothesis testing. For now you can implement the test statistic calculation based on this formula

<img width="278" alt="image" src="https://github.com/user-attachments/assets/074423ec-5459-49d4-9dbb-96afc68a3358" />

where n is the total sample size and R is rank.

**Comment by ashokkrish** [comment]
@sam-vb On the mainPanel for a K-W test let's have the following workflow

Main header: Hypothesis Test (bolded and underlined)

_H_0: The distributions of the groups are identical, and their medians are equal.
_H_a: At least one group differs in median from the others.

alpha = the value user selects from the radio button

**Test Statistic:**

Above formula = extract the TS-value from the `kruskal.wallis()` R function

Extract the p-value from the `kruskal.wallis()` R function.

Conclusion: 

Since the p-value is less than the significance level (actual p-value <= 𝛼), we reject the null hypothesis. There is sufficient evidence to conclude that at least one group's median is significantly different from the others.

or

Since the p-value is greater than the significance level (actual p-value > 𝛼), we fail to reject the null hypothesis. There is insufficient evidence to conclude that the medians of the groups are significantly different.

**Comment by sam-vb** [comment]
@ashokkrish I've just pushed the code with the main panel as requested above. Here is what it looks like:

![Image](https://github.com/user-attachments/assets/7261fc49-3ff2-41d5-8bbc-e32c872e961d)

Let me know if you find any issues or would like any changes/additions made.

**Comment by sam-vb** [comment]
@ashokkrish Updated with change to significance level:

![Image](https://github.com/user-attachments/assets/c6caac12-7c2e-4cde-ba57-e5ae983db2f0)

**Comment by ashokkrish** [comment]
Thanks, could you also round the test statistic (TS) value to for digits?

<img width="268" alt="Image" src="https://github.com/user-attachments/assets/881220a3-83ab-49a1-bb4c-c21d4271b29f" />

Display the answer as 10.6768.

**Comment by ashokkrish** [comment]
FUTURE WORK:

Reorder the workflow in the following manner

Let the user **Upload their Data**, next let them choose the **Data Format** after that let them choose the type of test

<img width="101" alt="Image" src="https://github.com/user-attachments/assets/4c81f7bc-ef65-46f8-b3a7-564a5f3eba62" />

Followed by

<img width="127" alt="Image" src="https://github.com/user-attachments/assets/65284cd2-3ed7-47a3-b1b9-5623f86f87e1" />

Followed by (only for ANOVA for now)

<img width="386" alt="Image" src="https://github.com/user-attachments/assets/a552b8a2-dd57-4ddd-9929-749fc2485488" />

This way the **data upload happens only once** and the user is able to run both ANOVA and KW test by a radio button toggle (either reactively or with the click of the Calculate button).

**Comment by sam-vb** [comment]
Fixed!

![Image](https://github.com/user-attachments/assets/7bab00c3-5736-42a5-bac0-85cd769a8ed2)

![Image](https://github.com/user-attachments/assets/671d5485-9995-4211-99e3-3b7f23545750)

**Comment by ashokkrish** [comment]
When running a ANOVA F test for a uploaded data on the mainPanel we have the following tabs

<img width="239" alt="Image" src="https://github.com/user-attachments/assets/cf6bd732-19b7-40c3-a77d-85567551fd0b" />

For a K-W test I would like two tabs with the first tab labelled Analysis and the second tab labelled uploaded data.

In the **Analysis** tab let's print the current flow on output

In the **Uploaded Data** tab let's print the uploaded data.

**Comment by ashokkrish** [comment]
> Fixed!

Could we please round the P-value to 4 digits as well?

**Comment by sam-vb** [comment]
Apologies, I misunderstood your earlier direction. P-value rounds to 4 decimal places now, and I've added the requested text above. Before I push, I want to clarify:

![Image](https://github.com/user-attachments/assets/5882a9c7-0c85-49e8-8b26-0f29caca2bb9)
Above is the current appearance of the KW main panel. Are there any changes you'd like to make?

Additionally, the p-value section for a chi-squared test has additional text about which hypothesis to accept/reject:

![Image](https://github.com/user-attachments/assets/58b8a814-51d7-4db2-a1ec-a2f9abd9cad7)

Would you like that added to the KW p-value? I am assuming we want to keep the conclusion as is for now.

**Comment by ashokkrish** [comment]
> Would you like that added to the KW p-value? I am assuming we want to keep the conclusion as is for now.

@sam-vb Yes and yes. This looks good. You can push it to the main branch.

**Comment by sam-vb** [comment]
> When running a ANOVA F test for a uploaded data on the mainPanel we have the following tabs
> <img alt="Image" width="239" src="https://private-user-images.githubusercontent.com/410949/406806471-cf6bd732-19b7-40c3-a77d-85567551fd0b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwMzgzMzMsIm5iZiI6MTczODAzODAzMywicGF0aCI6Ii80MTA5NDkvNDA2ODA2NDcxLWNmNmJkNzMyLTE5YjctNDBjMy1hNzdkLTg1NTY3NTUxZmQwYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyOFQwNDIwMzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NzJkZTQ1ZjNhZjdmNmU1ZjFmZTg0YjRjY2Q4ODM3OTAyYzFmZDY3NGJjMjk4NzkxYmMxODRmMjAxMGJhMDRjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.LMXS4qlAxceqOKnovpPq1-ahz4R7jE9P8xmRGAhZ8hg">
> 
> For a K-W test I would like two tabs with the first tab labelled Analysis and the second tab labelled uploaded data.
> 
> In the **Analysis** tab let's print the current flow on output
> 
> In the **Uploaded Data** tab let's print the uploaded data.

Changes have been pushed. I can work on these tabs next ('Graphs' tab omitted) and complete by Wednesday to leave room for deployment by Friday. Let me know of any other changes/tasks you'd like completed by then.

**Comment by ashokkrish** [comment]
@sam-vb 

<img width="179" alt="Image" src="https://github.com/user-attachments/assets/61c82b6a-2269-4879-8801-b837034704b7" />

I would like the x2 be replaced by the Greek letter (Chi-square). This symbol was previously used in 

<img width="249" alt="Image" src="https://github.com/user-attachments/assets/49f1b2ba-376e-4026-b292-cb74cce6fcbc" />

Would appreciate if you can fix this please.

**Comment by ashokkrish** [comment]
> I can work on these tabs next ('Graphs' tab omitted) and complete by Wednesday to leave room for deployment by Friday.

@sam-vb Yes, please proceed with this tasks and periodically push the code to the main branch so I can test it out.

**Comment by sam-vb** [comment]
Added tabs. Note that the appearance of "Hypothesis Testing" title on the Analysis tab when bolded and underlined as requested is inconsistent with the "Data File" title on the Uploaded Data tab. Per ANOVA tabs, all titles of all tabs are neither bolded or underlined - let me know what you want for KW.

**Comment by ashokkrish** [comment]
@sam-vb I have a minor request in K-W test conclusion. Currently it says  (actual p-value ≤α) what I meant here is to grab the actual p-value and the user selected alpha and put them in the brackets (0.0048≤0.05) for the Bone Densities dataset.

**Comment by sam-vb** [comment]
Yes, for sure. 

![Image](https://github.com/user-attachments/assets/0be22b22-e5ea-48a7-a44d-fb47628b395f)

Are there any other formats you'd like changed?

**Comment by ashokkrish** [comment]
This is perfect. We can close this issue now. I will have Bryce deploy this weekend.

**Comment by ashokkrish** [comment]
@dianahuff The conclusion if swapped

<img width="405" alt="Image" src="https://github.com/user-attachments/assets/c988cc68-775a-4361-9282-dd4af9833734" />

In this case the null hypothesis is not rejected so there isn't enough statistical evidence to support the alternative hypothesis.
